### PR TITLE
Rust: extract parse errors as diagnostics

### PR DESCRIPTION
**Pull request by aibaars**
Depends on https://github.com/github/codeql/pull/17543

This pull request add extraction of parse error messages, and adds some diagnostic queries.

### CODE REVIEW

File: rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected [hunk 1]
```diff
@@ -0,0 +1,6 @@
+| does_not_compile.rs:2:6:2:5 | expected SEMICOLON | Extraction failed in does_not_compile.rs with error expected SEMICOLON | 2 |
+| does_not_compile.rs:2:9:2:8 | expected SEMICOLON | Extraction failed in does_not_compile.rs with error expected SEMICOLON | 2 |
+| does_not_compile.rs:2:13:2:12 | expected SEMICOLON | Extraction failed in does_not_compile.rs with error expected SEMICOLON | 2 |
+| does_not_compile.rs:2:21:2:20 | expected SEMICOLON | Extraction failed in does_not_compile.rs with error expected SEMICOLON | 2 |
+| does_not_compile.rs:2:26:2:25 | expected SEMICOLON | Extraction failed in does_not_compile.rs with error expected SEMICOLON | 2 |
+| does_not_compile.rs:2:32:2:31 | expected field name or number | Extraction failed in does_not_compile.rs with error expected field name or number | 2 |
```
Comment (geoffw0) [inline rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected:6]: I'm surprised we don't get any results from `error.rs` (which contains `compile_error!("An error!");`).

File: rust/ql/src/queries/summary/NumberOfFilesExtractedWithErrors.ql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @id rust/summary/number-of-files-extracted-with-errors
+ * @name Total number of Rust files that were extracted with errors
+ * @description The total number of Rust code files that we extracted, but where
+ *  at least one extraction error occurred in the process.
```
Comment (geoffw0) [inline rust/ql/src/queries/summary/NumberOfFilesExtractedWithErrors.ql:5]:
```suggestion
 * @description The total number of Rust files in the source code directory that
 *  were extracted, but where at least one extraction error occurred in the process.
```

File: rust/ql/src/queries/summary/NumberOfSuccessfullyExtractedFiles.ql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @id rust/summary/number-of-successfully-extracted-files
+ * @name Total number of Rust files that were extracted without error
+ * @description The total number of Rust code files that we extracted without
+ *   encountering any extraction errors
```
Comment (geoffw0) [inline rust/ql/src/queries/summary/NumberOfSuccessfullyExtractedFiles.ql:5]:
```suggestion
 * @description The total number of Rust files in the source code directory that
 *  were extracted without encountering any extraction errors.
```

File: rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected (hunk 1 above)
Comment (aibaars) [reply rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected:6]: The line in error.rs isn't a syntax error.

Comment (geoffw0) [reply rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected:6]: I suppose.  I feel like it should show up somewhere (and is a cleaner test than `does_not_compile.rs`).

Comment (aibaars) [reply rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected:6]: The extractor currently only parses the code, and `error.rs` does not contain a syntax error. 
I would expect an error to be reported once we implement macro expansion.

Comment (geoffw0) [reply rust/ql/test/query-tests/diagnostics/ExtractionErrors.expected:6]: Sounds good to me. :+1:

File: rust/extractor/src/main.rs [hunk 4]
```diff
@@ -1,13 +1,48 @@
...
+        let valid = chunk.valid();
+        let invalid = chunk.invalid();
+        if invalid.is_empty() {
+            debug_assert_eq!(valid.len(), v.len());
+            return (Cow::Borrowed(valid), None);
+        }
+        (valid, invalid)
+    } else {
+        return (Cow::Borrowed(""), None);
+    };
```
Comment (geoffw0) [inline rust/extractor/src/main.rs:25]: Is the special handling of the first chunk essentially an optimisation for the common case?

**Review by geoffw0 — COMMENTED** [review]
Additional commits LGTM as well.  We're just waiting for https://github.com/github/codeql/pull/17543 I think.

File: rust/extractor/src/main.rs (hunk 4 above)
Comment (aibaars) [reply rust/extractor/src/main.rs:25]: No, it is a way to get the location of the first UTF8 decoding error, if any. If a file is complete garbage we don't want to generate a ton of useless decoding errors, one is enough.

Comment (aibaars) [reply rust/extractor/src/main.rs:25]: Although the `Cow` (copy-on-write) stuff is an optimisation for the common case that the input is entirely valid utf8 and no replacements are needed.

**Review by geoffw0 — APPROVED** [review]
LGTM.

**Pull request merged**